src/Simplex_tree/include/gudhi/Simplex_tree.h

```diff
@@ -1096,69 +1135,138 @@ class Simplex_tree {
     copy.clear();
     copy.insert(copy.end(), first, last);
     std::sort(copy.begin(), copy.end());
     auto last_unique = std::unique(copy.begin(), copy.end());
     copy.erase(last_unique, copy.end());
-    GUDHI_CHECK_code(
-      for (Vertex_handle v : copy)
-        GUDHI_CHECK(v != null_vertex(), "cannot use the dummy null_vertex() as a real vertex");
-    )
+    GUDHI_CHECK_code(for (Vertex_handle v : copy)
+                         GUDHI_CHECK(v != null_vertex(), "cannot use the dummy null_vertex() as a real vertex"););
     // Update dimension if needed. We could wait to see if the insertion succeeds, but I doubt there is much to gain.
     dimension_ = (std::max)(dimension_, static_cast<int>(std::distance(copy.begin(), copy.end())) - 1);
 
-    return rec_insert_simplex_and_subfaces_sorted(root(), copy.begin(), copy.end(), filtration);
+    if constexpr (Options::store_filtration){
+      if (insertion_strategy == Insertion_strategy::LOWEST)
+        return _rec_insert_simplex_and_subfaces_sorted(root(), copy.begin(), copy.end(), filtration);
+      else if (insertion_strategy == Insertion_strategy::HIGHEST)
+        return _insert_simplex_and_subfaces_at_highest(root(), copy.begin(), copy.end(), filtration);
+      else
+        return _insert_simplex_and_subfaces_at_highest(root(), copy.begin(), copy.end(), 0);
+    } else {
+      return _rec_insert_simplex_and_subfaces_sorted(root(), copy.begin(), copy.end(), filtration);
+    }
+    
   }
 
  private:
   // To insert {1,2,3,4}, we insert {2,3,4} twice, once at the root, and once below 1.
-  template <class ForwardVertexIterator>
-  std::pair<Simplex_handle, bool> rec_insert_simplex_and_subfaces_sorted(Siblings* sib,
-                                                                         ForwardVertexIterator first,
-                                                                         ForwardVertexIterator last,
-                                                                         const Filtration_value& filt)
-  {
+  template <class ForwardVertexIterator, bool update_fil = true>
+  std::pair<Simplex_handle, bool> _rec_insert_simplex_and_subfaces_sorted(Siblings* sib,
+                                                                          ForwardVertexIterator first,
+                                                                          ForwardVertexIterator last,
+                                                                          const Filtration_value& filt) {
     // An alternative strategy would be:
     // - try to find the complete simplex, if found (and low filtration) exit
     // - insert all the vertices at once in sib
     // - loop over those (new or not) simplices, with a recursive call(++first, last)
     Vertex_handle vertex_one = *first;
 
-    if (++first == last) return insert_node_<true, false, true>(sib, vertex_one, filt);
+    if (++first == last) return insert_node_<update_fil, false, update_fil>(sib, vertex_one, filt);
```

Comment:
A small comment would be welcome here, I was lost with the templates arguments :sweat_smile: 
(just like the one below where you just say `we are building the whole subtree from scratch`)

Reply:
I added a comment explaining the template arguments above. Is it okay like this?